### PR TITLE
Add guidelines for multi-api-version support

### PR DESCRIPTION
**Pull request by mikekistler**
This PR attempts to capture the general guidelines and recommended design for multi-api-version support in Azure SDKs.

### CODE REVIEW

File: docs/general/design.md [hunk 1]
```diff
@@ -109,6 +109,22 @@ The purposes of the client library is to communicate with an Azure service.  Azu
 
 {% include requirement/MUST id="general-service-apiversion-7" %} add or replace the `api-version` query parameter on any URI returned by the service e.g., `Operation-Location`, next page links, etc., with the service version passed configured on the client.
 
+{% include requirement/MUST id="general-service-apiversion-8" %} validate that query and header parameters passed to a method are defined for the api-version specified when the client was created and return an error if the validation fails.
```
Comment (lmazuel) [inline docs/general/design.md:112]: Nit: id `8` is already used line 98 for some reason

Comment (lmazuel) [inline docs/general/design.md:112]: We should be explicit about the content of error. Something like: The error should specify the current api-version and the possible set of api-versions available for this option.

As the version range is available in the TypeSpec, making the error message show the possible range should be do-able, and would bring clarity to customers on how to recover and update their code.

Comment (weidongxu-microsoft) [reply docs/general/design.md:112]: This may not be easy to do in Swagger, as it is hard for codegen to know "when" the query/header was added.

TypeSpec would be OK.

File: docs/general/design.md [hunk 2]
```diff
@@ -109,6 +109,22 @@ The purposes of the client library is to communicate with an Azure service.  Azu
...
+
+Validation must be performed before sending the request to the service to avoid unexpected behavior in the event the service does not reject a request with an unrecognized header or query parameter.
+
+{% include requirement/MUST id="general-service-apiversion-9" %} define a single class/struct for all compatible versions of a model type.
+
+Azure's versioning rules generally only allow properties to be added -- not removed, renamed, or modified in an incompatible way. In this (common) case, the latest version of the model is compatible with all earlier versions, and using this to represent all versions simplifies the API without loss of functionality.
```
Comment (lirenhe) [inline docs/general/design.md:118]: Shall we seperate preview versions and stable versions here?

File: docs/general/design.md (hunk 1 above)
Comment (mikekistler) [reply docs/general/design.md:112]: Thanks for catching that! I assumed these were sequential.  Silly me! 😜

Comment (mikekistler) [reply docs/general/design.md:112]: Good suggestion but I think it should be sufficient for the error to specify the earliest api-version that defines the parameter.  I'll add that.

File: docs/general/design.md (hunk 2 above)
Comment (mikekistler) [reply docs/general/design.md:118]: I agree that we should clarify how the rules apply for previews. I'll add that.

**Comment by mikekistler** [comment]
Adding the architects and PMS as reviewers.  Please review and leave comments/questions or an approval.

File: docs/general/design.md [hunk 3]
```diff
@@ -109,6 +109,24 @@ The purposes of the client library is to communicate with an Azure service.  Azu
...
+
+Note: Azure versioning rules allow a new preview or GA api-version to be breaking in relation to a prior preview, but this isn't a concern since any libraries for the earlier preview should be beta versions which are exempt from compatibility guarantees.
+
+{% include requirement/MAY id="general-service-apiversion-11" %} validate that request body properties passed to a method are defined for the api-version specified when the client was created.
+
+Validating request body properties is not required because the cost in terms of code complexity and runtime performance is prohibitive in most languages.
+
+{% include requirement/MUSTNOT id="general-service-apiversion-12" %} validate that response body properties returned by the service are defined for the api-version specified on the request.
```
[the remark this answers is not in the document]
Comment (johanste) [reply docs/general/design.md:126]: Agreed. This feels like we are doing work that only makes things worse for the caller. We have already disabled required checks pretty much across the board because doing so turned an annoyance into a a true problem.

File: docs/general/design.md [hunk 4]
```diff
@@ -109,6 +109,24 @@ The purposes of the client library is to communicate with an Azure service.  Azu
 
 {% include requirement/MUST id="general-service-apiversion-7" %} add or replace the `api-version` query parameter on any URI returned by the service e.g., `Operation-Location`, next page links, etc., with the service version passed configured on the client.
 
+{% include requirement/MUST id="general-service-apiversion-9" %} validate that query and header parameters passed to a method are defined for the api-version specified when the client was created and return an error if the validation fails.
+
+Validation must be performed before sending the request to the service to avoid unexpected behavior in the event the service does not reject a request with an unrecognized header or query parameter. When the validation fails, the error returned to the client should indicate the first api-version that allows the parameter to be specified.
+
+{% include requirement/MUST id="general-service-apiversion-10" %} define a single class/struct for all compatible versions of a model type.
```
Comment (johanste) [inline docs/general/design.md:116]: This effectively tell languages that have richer type systems that they are not allowed to use said type system. For example, Python could provide different protocol types/have per API version typing (we don't in latest DPG, but we did for control plane libraries). So I would leave this out of the requirements.

File: docs/general/design.md [hunk 5]
```diff
@@ -109,6 +109,24 @@ The purposes of the client library is to communicate with an Azure service.  Azu
 
 {% include requirement/MUST id="general-service-apiversion-7" %} add or replace the `api-version` query parameter on any URI returned by the service e.g., `Operation-Location`, next page links, etc., with the service version passed configured on the client.
 
+{% include requirement/MUST id="general-service-apiversion-9" %} validate that query and header parameters passed to a method are defined for the api-version specified when the client was created and return an error if the validation fails.
+
+Validation must be performed before sending the request to the service to avoid unexpected behavior in the event the service does not reject a request with an unrecognized header or query parameter. When the validation fails, the error returned to the client should indicate the first api-version that allows the parameter to be specified.
```
[the remark this answers is not in the document]
Comment (johanste) [reply docs/general/design.md:114]: Services are expected to ignore unknown headers (per HTTP).

File: docs/general/design.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (mikekistler) [reply docs/general/design.md:126]: The requirement here is "MUSTNOT" --
> YOU MUST NOT validate that response body properties returned by the service are defined for the api-version specified on the request.

@JeffreyRichter @johanste perhaps you thought it was MUST? Your comments suggest that you believe we should not validate, so I think you are agreeing with the guideline as written.

File: docs/general/design.md [hunk 6]
```diff
@@ -109,6 +109,24 @@ The purposes of the client library is to communicate with an Azure service.  Azu
...
+
+{% include requirement/MUST id="general-service-apiversion-10" %} define a single class/struct for all compatible versions of a model type.
+
+Azure's versioning rules generally only allow properties to be added -- not removed, renamed, or modified in an incompatible way. In this (common) case, the latest version of the model is compatible with all earlier versions, and using this to represent all versions simplifies the API without loss of functionality.
+
+Note: Azure versioning rules allow a new preview or GA api-version to be breaking in relation to a prior preview, but this isn't a concern since any libraries for the earlier preview should be beta versions which are exempt from compatibility guarantees.
+
+{% include requirement/MAY id="general-service-apiversion-11" %} validate that request body properties passed to a method are defined for the api-version specified when the client was created.
```
[the remark this answers is not in the document]
Comment (johanste) [reply docs/general/design.md:122]: `MAY` guidelines are often not that helpful. At the very least, it should be a `SHOULD` or `SHOULD NOT` if we actually have a suggestion/opinion on what should be done. If we can't say either, we should not include it because as a SDK developer, I actually don't know what to do with it.

File: docs/general/design.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (johanste) [reply docs/general/design.md:126]: Same.

File: docs/general/design.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (mikekistler) [reply docs/general/design.md:122]: Then I'd be inclined to make this a `SHOULD NOT`.

File: docs/general/design.md (hunk 4 above)
Comment (mikekistler) [reply docs/general/design.md:116]: That was not my intent. The intent was to guide SDK developers to avoid creating many "mostly duplicate" classes/structs that only differ in minor and generally compatible details. So if V2 just adds a new property, make a single class/struct that has all the V1 properties and this new V2 property. When used with a V1 api-version the V2 property should be empty on requests and not set on responses. I think it will be overly burdensome on users if they must choose amongst n subtly different classes for a service object, and we want to avoid that where possible.

On the other hand, if v2 changes some field type from int32 to int64, then maybe that's not something that can be properly represented with just one class/struct and then the library can contain two different class/struct definitions. But then it should also provide guidance on when to use one vs the other.

Comment (bterlson) [reply docs/general/design.md:116]: I don't the concern of "mostly duplicate" applies to TypeScript where you generally don't reference a type by name in code (so less confusion concerns) and separation helps the type system validate everything at tooling time rather than waiting for runtime validation.

File: docs/general/design.md [hunk 7]
```diff
@@ -109,6 +109,28 @@ The purposes of the client library is to communicate with an Azure service.  Azu
...
 
+{% include requirement/MUST id="general-service-apiversion-9" %} validate that query and header parameters passed to a method are defined for the api-version specified when the client was created and return an error if the validation fails.
+
+Validation must be performed before sending the request to the service to avoid unexpected behavior in the event the service does not reject a request with an unrecognized header or query parameter. When the validation fails, the error returned to the client should indicate the first api-version that allows the parameter to be specified.
+
+{% include requirement/MUST id="general-service-apiversion-10" %} define a single class/struct for all compatible versions of a model type.
+
+Azure's versioning rules generally only allow properties to be added -- not removed, renamed, or modified in an incompatible way. In this (common) case, the latest version of the model is compatible with all earlier versions, and using this to represent all versions simplifies the API without loss of functionality.
```
Comment (witemple-msft) [inline docs/general/design.md:118]: Isn't this basically only true (a) for **output only** models, where input models adding properties is only non-breaking if those properties are optional; and (b) in contexts where properties are implicitly optional? In TypeScript, it's very common to enable strict optionality, so if a property is added to an output model and it's required, we can't just use that model for older API versions where that property is not returned. It makes the type incorrect.

This echoes a concern that Brian was talking about in the previous point: we have an issue with multi-api-version modeling where when we try to account for all possible API versions, we end up with a type surface that is "infinitely loosening" until we decide to accept a breaking change that will tighten the models by removing support for older API versions.

This has happened in Form Recognizer, for example, where new required output properties were introduced that we cannot model in that hand-written HLC without _lots_ of bespoke code that we aren't willing to maintain, so instead we just make the property optional. In general we can't just assume that the generated models for the newest API version are compatible with the older versions for this reason.

File: docs/general/design.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (witemple-msft) [reply docs/general/design.md:114]: I guess I don't understand why we MUST do this for query/header params but SHOULD NOT do so for body props.

File: docs/general/design.md (hunk 4 above)
Comment (witemple-msft) [reply docs/general/design.md:116]: I think we really have to avoid the problem of "infinitely loosening" type surface in the TS packages (talked about this a little more in the comment on the below guideline). With strict undefined/optionality checking in the TypeScript compiler, it _really_ sucks to have properties that are mandatory on the API version you're actually using but optional on the type because of API compatibility. The user ends up having to do a bunch of defensive checking against infallible cases just to make the compiler happy.

Comment (johanste) [reply docs/general/design.md:116]: My SEK 0,02 - I do not approve of this guideline. This is a per-language decision and it has no place in the general guidelines. If we start adding these kind of requirements, we are going to end up with the lowest common denominator for all languages. I would not blindly apply this to Python. There are way too many trade-offs/scenario specific info to take into account.

File: docs/general/design.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (johanste) [reply docs/general/design.md:114]: For header parameters, it makes some sense because the service is explicitly expected to silently ignore them. For query parameters, it depends - in general, they are not expected to be ignored, and thus the service can/should reject the call.

I honestly think that these guidelines are already covered in the client validation section, and this addition introduces as many questions as it answers.

File: docs/general/design.md (hunk 4 above)
Comment (mikekistler) [reply docs/general/design.md:116]: Okay ... I will remove this.

File: docs/general/implementation.md [hunk 8]
```diff
@@ -65,6 +65,12 @@ The service client will have methods that send requests to the service. These me
...
 
+### Validation in multi-api-version clients
+
+A service client that supports multiple service versions (see [Service API Versions]) may expose parameters, models, and methods that are not supported by all service versions. While the above guidance generally applies for these situations, some special provisions are appropriate. An example is a service that adds support for conditional requests in a new service version. To support this, the client will expose parameters for the conditional request headers ("if-match", "if-unmodified-since", etc), but if these are used with a client configured with an earlier service version the result is potentially dangerous. If the service does not check for and reject requests that specify these headers with the earlier api-version, if may perform operations that violate the condition.
+
+{% include requirement/MUST id="general-params-apiversion-9" %} validate that query and header parameters passed to a method are defined for the api-version specified when the client was constructed and return an error if the validation fails.
```
Comment (johanste) [inline docs/general/implementation.md:72]: This is still only true for things that the service doesn't already validate. Which is what we (at least try to) say w. regards to client side validation.

It also would only apply to parameters (query or header parameters) that the client knows have been added in the future, but are not supported by the current API version. For example, we have parameters that are intentionally open and allows callers to provide custom headers - and supporting this is a requirement (e.g. to pass through proxies etc.)

Are we in a position where the example given is the only time that headers should be validated?

Comment (mikekistler) [reply docs/general/implementation.md:72]: > It also would only apply to parameters (query or header parameters) that the client knows have been added in the future, but are not supported by the current API version.

I agree that this is the scope and nothing broader than this. The ideas is that in a new api-version there is some new header or query parameter "foo". That parameter is made available for users, probably through a parameter or property in an options bag. The client should check that the user has set the api-version appropriately before processing this new parameter.

The rationale here is that the api-version is set when the client is constructed, and in an enterprise application that might be far away from where the client is used to make calls to the service. So a developer _may not know_ which api-version was set -- may not even think to look. What the developer knows is that intellisense and the docs are offering this parameter that looks like what he wants. So he/she sets it and if the method succeeds has no indication that the parameter had no effect.

Do we want to check for this case or not?

Comment (johanste) [reply docs/general/implementation.md:72]: We do not want to check for the body parameter. We only want to check for the header parameter if it is something service specific significantly important that the service would have otherwise silently ignored.

Comment (mikekistler) [reply docs/general/implementation.md:72]: I agree that we do not want to check for body parameters and this is why they are not included in the guideline.

How would you suggest rewording the guideline to address the rest of your comment?
